wp-content/plugins/core/src/Templates/Components/Component.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+namespace Tribe\Project\Templates\Components;
+
+use Tribe\Libs\Utils\Markup_Utils;
+use DI;
+use Tribe\Project\Assets\Theme\Theme_Build_Parser;
+use Tribe\Project\Components\Component_Factory;
+use Tribe\Project\Components\Handler;
+use Twig\Environment;
+use Twig\Error\Error;
+
+abstract class Component {
+
+	/**
+	 * @var array
+	 */
+	protected $data;
+
+	/**
+	 * @var Component_Factory
+	 */
+	protected $factory;
+
+	/**
+	 * @var Environment
+	 */
+	protected $twig;
+
+	/**
+	 * @var Theme_Build_Parser
+	 */
+	protected $build_parser;
+
+	public function __construct( Component_Factory $factory, Environment $twig, Theme_Build_Parser $build_parser, $args = [] ) {
+		$this->factory      = $factory;
+		$this->twig         = $twig;
+		$this->build_parser = $build_parser;
+
+		$this->data = $args;
+		$this->merge_defaults();
+	}
+
+	private function merge_defaults() {
+		$this->data = wp_parse_args( $this->data, $this->defaults() );
+	}
+
+	public function data(): array {
+		return $this->data;
+	}
+
+	public function output(): void {
+		/**
+		 * @var Component $component
+		 */
+		$this->init();
+
+		ob_start();
+		$this->render();
+		$template = ob_get_clean();
+
+		try {
+			echo $this->twig->render( $this->twig->createTemplate( $template ), $this->data() );
+		} catch ( Error $e ) {
+			error_log( $e->getMessage() );
+
+			return;
+		}
+	}
+
+	public function get_rendered_output(): string {
+		ob_start();
+		$this->output();
+
+		return ob_get_clean();
+	}
+
+	/**
+	 * Merge associative arrays of attributes into a string
+	 *
+	 * @param array ...$attributes
+	 *
+	 * @return string
+	 */
+	protected function merge_attrs( array ...$attributes ): string {
```

Comment:
If I understand the change in [this commit](https://github.com/moderntribe/square-one/pull/501/commits/37b8d1605f5932a95b5ab6fb1a6d2dd347fc403d), these two methods (`merge_attrs` & `merge_classes`) are no longer needed, correct?

Reply:
You do, and they are! Great catch.

Reply:
Woot! I love the Twig filters for this purpose, btw. It's a really nice enhancement.